File: cloud-web-app/web/lib/server/quant/risk-envelope.ts
/**
 * N5 — Risk envelope (web mirror of Rust kernel gates).
 * Fail-closed: live trading never enabled. Pairs with N3 trade audit.
 * Canonical reject-before-network lives in `aethel-kernel-rust::risk_envelope`.
 * Distinct from Hub Coins / Creative CostGuard.
 */

import { createComponentLogger } from '@/lib/observability/logger'

const log = createComponentLogger('risk-envelope')

/** Hard-disabled — investment-grade HELD. */
export const LIVE_TRADING_ENABLED = false as const

export const DEFAULT_MAX_DRAWDOWN_BPS = 1000
export const DEFAULT_MAX_LEVERAGE_X100 = 200
export const DEFAULT_MAX_NOTIONAL_USD = 100_000

export type RiskRejectCode =
  | 'kill_switch'
  | 'live_trading_disabled'
  | 'drawdown_exceeded'
  | 'leverage_exceeded'
  | 'notional_exceeded'
  | 'invalid_request'

export interface RiskEnvelopeLimits {
  maxDrawdownBps: number
  maxLeverageX100: number
  maxNotionalUsd: number
  killSwitchArmed: boolean
  liveTradingEnabled: typeof LIVE_TRADING_ENABLED
}

export interface RiskCheckRequest {
  strategyId: string
  notionalUsd: number
  leverageX100: number
  currentDrawdownBps: number
  wantsLive: boolean
}

export type RiskVerdict =
  | { ok: true }
  | { ok: false; code: RiskRejectCode; reason: string }

export function createRiskEnvelopeLimits(
  overrides?: Partial<Omit<RiskEnvelopeLimits, 'liveTradingEnabled'>>,
): RiskEnvelopeLimits {
  return {
    maxDrawdownBps: overrides?.maxDrawdownBps ?? DEFAULT_MAX_DRAWDOWN_BPS,
    maxLeverageX100: overrides?.maxLeverageX100 ?? DEFAULT_MAX_LEVERAGE_X100,
    maxNotionalUsd: overrides?.maxNotionalUsd ?? DEFAULT_MAX_NOTIONAL_USD,
    killSwitchArmed: overrides?.killSwitchArmed ?? false,
    liveTradingEnabled: LIVE_TRADING_ENABLED,
  }
}

export function armKillSwitch(limits: RiskEnvelopeLimits): RiskEnvelopeLimits {
  return { ...limits, killSwitchArmed: true }
}

export function disarmKillSwitch(limits: RiskEnvelopeLimits): RiskEnvelopeLimits {
  return { ...limits, killSwitchArmed: false, liveTradingEnabled: LIVE_TRADING_ENABLED }
}

/** Evaluate order against envelope — reject before broker/network. */
export function evaluateRisk(
  limits: RiskEnvelopeLimits,
  req: RiskCheckRequest,
): RiskVerdict {
  if (
    !req.strategyId.trim() ||
    !Number.isFinite(req.notionalUsd) ||
    req.notionalUsd <= 0
  ) {
    return { ok: false, code: 'invalid_request', reason: 'strategyId empty or notional invalid' }
  }

  if (limits.killSwitchArmed) {
    return { ok: false, code: 'kill_switch', reason: 'kill-switch armed — all orders rejected' }
  }

  if (req.wantsLive || limits.liveTradingEnabled) {
    return {
      ok: false,
      code: 'live_trading_disabled',
      reason: 'live trading disabled — paper/sandbox only',
    }
  }

  if (req.currentDrawdownBps > limits.maxDrawdownBps) {
    return {
      ok: false,
      code: 'drawdown_exceeded',
      reason: `drawdown ${req.currentDrawdownBps} bps exceeds max ${limits.maxDrawdownBps} bps`,
    }
  }

  if (req.leverageX100 > limits.maxLeverageX100) {
    return {
      ok: false,
      code: 'leverage_exceeded',
      reason: `leverage ${req.leverageX100}x100 exceeds max ${limits.maxLeverageX100}x100`,
    }
  }

  if (req.notionalUsd > limits.maxNotionalUsd) {
    return {
      ok: false,
      code: 'notional_exceeded',
      reason: `notional ${req.notionalUsd} exceeds max ${limits.maxNotionalUsd}`,
    }
  }

  return { ok: true }
}

export type RiskEnvelopeProbeResult = {
  ready: boolean
  liveTradingEnabled: false
  investmentGrade: false
  killSwitchRejects: boolean
  liveIntentRejects: boolean
  drawdownGateWorks: boolean
  leverageGateWorks: boolean
  paperPassWithinLimits: boolean
  path: string
  rustPath: string
  note: string
}

/** Probe N5 readiness — web mirror + documented Rust kernel path. */
export function probeRiskEnvelopeReadiness(): RiskEnvelopeProbeResult {
  let limits = createRiskEnvelopeLimits()

  const paperPassWithinLimits = evaluateRisk(limits, {
    strategyId: 'probe-paper',
    notionalUsd: 1000,
    leverageX100: 100,
    currentDrawdownBps: 50,
    wantsLive: false,
  }).ok

  const liveIntentRejects =
    evaluateRisk(limits, {
      strategyId: 'probe-live',
      notionalUsd: 1000,
      leverageX100: 100,
      currentDrawdownBps: 0,
      wantsLive: true,
    }).ok === false

  const drawdownGateWorks =
    evaluateRisk(limits, {
      strategyId: 'probe-dd',
      notionalUsd: 1000,
      leverageX100: 100,
      currentDrawdownBps: limits.maxDrawdownBps + 1,
      wantsLive: false,
    }).ok === false

  const leverageGateWorks =
    evaluateRisk(limits, {
      strategyId: 'probe-lev',
      notionalUsd: 1000,
      leverageX100: limits.maxLeverageX100 + 1,
      currentDrawdownBps: 0,
      wantsLive: false,
    }).ok === false

  limits = armKillSwitch(limits)
  const killSwitchRejects =
    evaluateRisk(limits, {
      strategyId: 'probe-kill',
      notionalUsd: 100,
      leverageX100: 100,
      currentDrawdownBps: 0,
      wantsLive: false,
    }).ok === false

  const ready =
    paperPassWithinLimits &&
    liveIntentRejects &&
    drawdownGateWorks &&
    leverageGateWorks &&
    killSwitchRejects &&
    !LIVE_TRADING_ENABLED

  log.info('risk_envelope_probed', {
    ready,
    liveTradingEnabled: LIVE_TRADING_ENABLED,
    investmentGrade: false,
  })

  return {
    ready,
    liveTradingEnabled: false,
    investmentGrade: false,
    killSwitchRejects,
    liveIntentRejects,
    drawdownGateWorks,
    leverageGateWorks,
    paperPassWithinLimits,
    path: 'lib/server/quant/risk-envelope.ts',
    rustPath: 'packages/aethel-kernel-rust/src/risk_envelope.rs',
    note: ready
      ? 'N5 PARTIAL — drawdown/leverage/kill-switch fail-closed; live hard-disabled; Rust kernel + web mirror; investmentGrade false'
      : 'N5 risk envelope probe failed',
  }
}

File: packages/aethel-kernel-rust/src/risk_envelope.rs
//! N5 — Rust risk envelope (max drawdown, leverage, kill-switch).
//! Fail-closed: live trading is never enabled. Kernel rejects before network.
//! Distinct from Hub Coins / Creative CostGuard. Pairs with N3 trade audit.

use serde::{Deserialize, Serialize};

/// Live trading is hard-disabled in this kernel surface (investment-grade HELD).
pub const LIVE_TRADING_ENABLED: bool = false;

/// Default max drawdown: 10% (1000 bps).
pub const DEFAULT_MAX_DRAWDOWN_BPS: u32 = 1000;

/// Default max leverage: 2.0× encoded as ×100.
pub const DEFAULT_MAX_LEVERAGE_X100: u32 = 200;

/// Default max order notional (paper sandbox ceiling).
pub const DEFAULT_MAX_NOTIONAL_USD: f64 = 100_000.0;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RiskRejectCode {
    KillSwitch,
    LiveTradingDisabled,
    DrawdownExceeded,
    LeverageExceeded,
    NotionalExceeded,
    InvalidRequest,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RiskEnvelopeLimits {
    pub max_drawdown_bps: u32,
    pub max_leverage_x100: u32,
    pub max_notional_usd: f64,
    /// When true, every order is rejected (fail-closed halt).
    pub kill_switch_armed: bool,
    /// Always false — never enable live trading from this envelope.
    pub live_trading_enabled: bool,
}

impl Default for RiskEnvelopeLimits {
    fn default() -> Self {
        Self {
            max_drawdown_bps: DEFAULT_MAX_DRAWDOWN_BPS,
            max_leverage_x100: DEFAULT_MAX_LEVERAGE_X100,
            max_notional_usd: DEFAULT_MAX_NOTIONAL_USD,
            kill_switch_armed: false,
            live_trading_enabled: LIVE_TRADING_ENABLED,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RiskCheckRequest {
    pub strategy_id: String,
    pub notional_usd: f64,
    pub leverage_x100: u32,
    pub current_drawdown_bps: u32,
    /// Caller intent — always rejected while live_trading_enabled is false.
    pub wants_live: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RiskVerdict {
    Pass,
    Reject {
        code: RiskRejectCode,
        reason: String,
    },
}

impl RiskVerdict {
    pub fn is_pass(&self) -> bool {
        matches!(self, RiskVerdict::Pass)
    }
}

/// Arm kill-switch — irreversible for this envelope instance until rebuilt.
pub fn arm_kill_switch(limits: &mut RiskEnvelopeLimits) {
    limits.kill_switch_armed = true;
}

/// Disarm only when not live (live remains hard-disabled).
pub fn disarm_kill_switch(limits: &mut RiskEnvelopeLimits) -> Result<(), RiskRejectCode> {
    if limits.live_trading_enabled {
        return Err(RiskRejectCode::LiveTradingDisabled);
    }
    limits.kill_switch_armed = false;
    Ok(())
}

/// Evaluate order against envelope — reject before any network/broker path.
pub fn evaluate_risk(limits: &RiskEnvelopeLimits, req: &RiskCheckRequest) -> RiskVerdict {
    if req.strategy_id.trim().is_empty()
        || !req.notional_usd.is_finite()
        || req.notional_usd <= 0.0
    {
        return RiskVerdict::Reject {
            code: RiskRejectCode::InvalidRequest,
            reason: "strategy_id empty or notional invalid".into(),
        };
    }

    if limits.kill_switch_armed {
        return RiskVerdict::Reject {
            code: RiskRejectCode::KillSwitch,
            reason: "kill-switch armed — all orders rejected".into(),
        };
    }

    // Fail-closed: live never ships from this kernel.
    if req.wants_live || limits.live_trading_enabled {
        return RiskVerdict::Reject {
            code: RiskRejectCode::LiveTradingDisabled,
            reason: "live trading disabled — paper/sandbox only".into(),
        };
    }

    if req.current_drawdown_bps > limits.max_drawdown_bps {
        return RiskVerdict::Reject {
            code: RiskRejectCode::DrawdownExceeded,
            reason: format!(
                "drawdown {} bps exceeds max {} bps",
                req.current_drawdown_bps, limits.max_drawdown_bps
            ),
        };
    }

    if req.leverage_x100 > limits.max_leverage_x100 {
        return RiskVerdict::Reject {
            code: RiskRejectCode::LeverageExceeded,
            reason: format!(
                "leverage {}x100 exceeds max {}x100",
                req.leverage_x100, limits.max_leverage_x100
            ),
        };
    }

    if req.notional_usd > limits.max_notional_usd {
        return RiskVerdict::Reject {
            code: RiskRejectCode::NotionalExceeded,
            reason: format!(
                "notional {} exceeds max {}",
                req.notional_usd, limits.max_notional_usd
            ),
        };
    }

    RiskVerdict::Pass
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RiskEnvelopeProbe {
    pub risk_envelope_ready: bool,
    pub live_trading_enabled: bool,
    pub investment_grade: bool,
    pub kill_switch_rejects: bool,
    pub live_intent_rejects: bool,
    pub drawdown_gate_works: bool,
    pub leverage_gate_works: bool,
    pub paper_pass_within_limits: bool,
    pub letter: String,
    pub note: String,
}

/// Soak-gated honesty probe — never flips investment_grade or live trading.
pub fn probe_risk_envelope() -> RiskEnvelopeProbe {
    let mut limits = RiskEnvelopeLimits::default();

    let paper_ok = evaluate_risk(
        &limits,
        &RiskCheckRequest {
            strategy_id: "probe-paper".into(),
            notional_usd: 1_000.0,
            leverage_x100: 100,
            current_drawdown_bps: 50,
            wants_live: false,
        },
    )
    .is_pass();

    let live_reject = matches!(
        evaluate_risk(
            &limits,
            &RiskCheckRequest {
                strategy_id: "probe-live".into(),
                notional_usd: 1_000.0,
                leverage_x100: 100,
                current_drawdown_bps: 0,
                wants_live: true,
            },
        ),
        RiskVerdict::Reject {
            code: RiskRejectCode::LiveTradingDisabled,
            ..
        }
    );

    let drawdown_reject = matches!(
        evaluate_risk(
            &limits,
            &RiskCheckRequest {
                strategy_id: "probe-dd".into(),
                notional_usd: 1_000.0,
                leverage_x100: 100,
                current_drawdown_bps: limits.max_drawdown_bps + 1,
                wants_live: false,
            },
        ),
        RiskVerdict::Reject {
            code: RiskRejectCode::DrawdownExceeded,
            ..
        }
    );

    let leverage_reject = matches!(
        evaluate_risk(
            &limits,
            &RiskCheckRequest {
                strategy_id: "probe-lev".into(),
                notional_usd: 1_000.0,
                leverage_x100: limits.max_leverage_x100 + 1,
                current_drawdown_bps: 0,
                wants_live: false,
            },
        ),
        RiskVerdict::Reject {
            code: RiskRejectCode::LeverageExceeded,
            ..
        }
    );

    arm_kill_switch(&mut limits);
    let kill_reject = matches!(
        evaluate_risk(
            &limits,
            &RiskCheckRequest {
                strategy_id: "probe-kill".into(),
                notional_usd: 100.0,
                leverage_x100: 100,
                current_drawdown_bps: 0,
                wants_live: false,
            },
        ),
        RiskVerdict::Reject {
            code: RiskRejectCode::KillSwitch,
            ..
        }
    );

    let ready = paper_ok
        && live_reject
        && drawdown_reject
        && leverage_reject
        && kill_reject
        && !LIVE_TRADING_ENABLED;

    RiskEnvelopeProbe {
        risk_envelope_ready: ready,
        live_trading_enabled: LIVE_TRADING_ENABLED,
        investment_grade: false,
        kill_switch_rejects: kill_reject,
        live_intent_rejects: live_reject,
        drawdown_gate_works: drawdown_reject,
        leverage_gate_works: leverage_reject,
        paper_pass_within_limits: paper_ok,
        letter: "n5".into(),
        note: if ready {
            "N5 risk envelope PARTIAL — drawdown/leverage/kill-switch fail-closed; live trading hard-disabled; investmentGrade false"
                .into()
        } else {
            "N5 risk envelope probe failed — gates incomplete".into()
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn paper_order_within_limits_passes() {
        let limits = RiskEnvelopeLimits::default();
        let v = evaluate_risk(
            &limits,
            &RiskCheckRequest {
                strategy_id: "s1".into(),
                notional_usd: 500.0,
                leverage_x100: 100,
                current_drawdown_bps: 10,
                wants_live: false,
            },
        );
        assert!(v.is_pass());
    }

    #[test]
    fn live_intent_always_rejected() {
        let limits = RiskEnvelopeLimits::default();
        let v = evaluate_risk(
            &limits,
            &RiskCheckRequest {
                strategy_id: "s1".into(),
                notional_usd: 100.0,
                leverage_x100: 100,
                current_drawdown_bps: 0,
                wants_live: true,
            },
        );
        assert!(matches!(
            v,
            RiskVerdict::Reject {
                code: RiskRejectCode::LiveTradingDisabled,
                ..
            }
        ));
        assert!(!LIVE_TRADING_ENABLED);
    }

    #[test]
    fn kill_switch_rejects_all() {
        let mut limits = RiskEnvelopeLimits::default();
        arm_kill_switch(&mut limits);
        let v = evaluate_risk(
            &limits,
            &RiskCheckRequest {
                strategy_id: "s1".into(),
                notional_usd: 50.0,
                leverage_x100: 100,
                current_drawdown_bps: 0,
                wants_live: false,
            },
        );
        assert!(matches!(
            v,
            RiskVerdict::Reject {
                code: RiskRejectCode::KillSwitch,
                ..
            }
        ));
    }

    #[test]
    fn drawdown_and_leverage_gates() {
        let limits = RiskEnvelopeLimits::default();
        assert!(matches!(
            evaluate_risk(
                &limits,
                &RiskCheckRequest {
                    strategy_id: "s1".into(),
                    notional_usd: 100.0,
                    leverage_x100: 100,
                    current_drawdown_bps: 10_001,
                    wants_live: false,
                },
            ),
            RiskVerdict::Reject {
                code: RiskRejectCode::DrawdownExceeded,
                ..
            }
        ));
        assert!(matches!(
            evaluate_risk(
                &limits,
                &RiskCheckRequest {
                    strategy_id: "s1".into(),
                    notional_usd: 100.0,
                    leverage_x100: 9999,
                    current_drawdown_bps: 0,
                    wants_live: false,
                },
            ),
            RiskVerdict::Reject {
                code: RiskRejectCode::LeverageExceeded,
                ..
            }
        ));
    }

    #[test]
    fn probe_ready_and_investment_grade_false() {
        let p = probe_risk_envelope();
        assert!(p.risk_envelope_ready);
        assert!(!p.live_trading_enabled);
        assert!(!p.investment_grade);
        assert_eq!(p.letter, "n5");
    }
}
